p2p/src/network/select/p2p_network_select_effects.rs

```diff
@@ -8,86 +13,39 @@ use crate::{
 
 use super::{super::*, p2p_network_select_state::P2pNetworkSelectStateInner, *};
 
-impl P2pNetworkSelectState {
-    pub fn incoming_data(
-        &self,
-        addr: ConnectionAddr,
-        kind: SelectKind,
-        data: Data,
-        fin: bool,
-        effects: &mut Vec<P2pNetworkSelectAction>,
-    ) {
-        fn forward_data_action(
-            addr: ConnectionAddr,
-            kind: SelectKind,
-            data: Data,
-            fin: bool,
-        ) -> P2pNetworkSelectAction {
-            match kind {
-                SelectKind::Authentication => P2pNetworkSelectAction::IncomingPayloadAuth {
-                    addr,
-                    fin,
-                    data: data.clone(),
-                },
-                SelectKind::Multiplexing(peer_id) => P2pNetworkSelectAction::IncomingPayloadMux {
-                    addr,
-                    peer_id: Some(peer_id),
-                    fin,
-                    data: data.clone(),
-                },
-                SelectKind::MultiplexingNoPeerId => P2pNetworkSelectAction::IncomingPayloadMux {
-                    addr,
-                    peer_id: None,
-                    fin,
-                    data: data.clone(),
-                },
-                SelectKind::Stream(peer_id, stream_id) => P2pNetworkSelectAction::IncomingPayload {
-                    addr,
-                    peer_id,
-                    stream_id,
-                    fin,
-                    data: data.clone(),
-                },
-            }
-        }
-        if matches!(&self.inner, P2pNetworkSelectStateInner::Error(..)) {
-            return;
-        }
-
-        if self.negotiated.is_some() {
-            effects.push(forward_data_action(addr, kind, data, fin));
-            return;
-        }
-
-        let payload_data = self.recv.buffer.clone();
+/// Propagates incoming data to corresponding action
+macro_rules! forward_incoming_data {
```

Comment:
This is up to you but in general we try to be conservative with the use of macros, and consider the pros and cons:

- Does the reduced code duplication/redundancy outweigh the increased complexity to understand and navigate the code?
- How can this affect coverage reports (profiling, fuzzing)?
- How can this affect debugging info/backtraces?

Reply:
I have tried putting it into new Action, but that triggers `NoRecursion` invariant, before there was function in state that putted all actions into vector and after that they were dispatched, this made debugging that part much harder. 

Problem is that state is immutable reference on store, but for dispatch store needs to be mutable, so I think pros outweigh cons.

Reply:
@0xMimir @0xMimir re: mutable access to the store, this is solved in the new style reducers (which combine reducer + dispatcher in the same handler function), but to take advantage of it the whole component needs to be ported: https://github.com/openmina/openmina/blob/898750c4007e8351e6d4792da234a20851253518/ARCHITECTURE.md#substate-access-queued-reducer-dispatch-and-callbacks

Probably not worth it just for this change but eventually we should have all our components ported